Errors and warnings for a form are now a property

JorisGoosen opened the pull request:
- No more need for directly manipulating QQuickItem*...
- Fixes https://github.com/jasp-stats/INTERNAL-jasp/issues/757
- Refactored some code a bit
- Added a property for Analysis * as well

JorisGoosen committed:
Errors and warnings for a form are now a property
- No more need for directly manipulating QQuickItem*...
- Fixes https://github.com/jasp-stats/INTERNAL-jasp/issues/757
- Refactored some code a bit
- Added a property for Analysis * as well

JorisGoosen commented on JASP-Desktop/analysis/analysisform.cpp:
```diff
@@ -721,49 +703,59 @@ void AnalysisForm::clearControlError(JASPControlBase* control)
-	if (!analysisVariant.isNull())
-	{
-		_analysis	= qobject_cast<Analysis *>(analysisVariant.value<QObject *>());
+	qmlContext(this)->setContextProperty("form", this);
```
Is this really necessary?

boutinb replied:
Yes: the form id is known is all forms, but not inside a JASP Widget. Like that it is known everywhere.

JorisGoosen replied:
But I keep seeing warnings in the output `QQmlContext: Cannot set property on internal context.`
I don't know if it is related to this though.

And also, I do not understand "the form id is known is all forms", what it would mean and how it is different for a JASP Widget?

boutinb replied:
I added this in this commit https://github.com/jasp-stats/jasp-desktop/commit/ffb7478a78660c76bf0c141a9c6b5cd60ab13953 to fix overflow issues in the form. I saw that sometimes the GridLayout did not have the form object, and I thought it was because GridLayout is not a JASPControl. But I have just tested it without using the context setting, and it still works. So you can remove it.

JorisGoosen replied:
Ok good!

JorisGoosen commented:
I'm also adding https://github.com/jasp-stats/INTERNAL-jasp/issues/1018 to this branch

JorisGoosen commented:
That is, if I get it to work. But it can be separated from this PR without problems if @boutinb is a quick man ;)

JorisGoosen committed:
No more javascript for errors on controls
- Form & Controls now check their children if those have errors
- Fixes https://github.com/jasp-stats/INTERNAL-jasp/issues/1018
- And children tell their parents about updates
- Only thing is that the little msgbox floating above a control that has an error is still instantiated from C++. I think this would be nicer with some kind of loader or so.

boutinb commented on JASP-Desktop/analysis/analysisform.cpp:
```diff
@@ -218,11 +218,11 @@ void AnalysisForm::addControl(JASPControlBase *control)
-			if (label.isEmpty())	_formErrorMessages.append(tr("Control with label '%1' has no name").arg(label));
-			else					_formErrorMessages.append(tr("A control has no name"));
+			if (label.isEmpty())	_formErrors.append(tr("Control with label '%1' has no name").arg(label));
+			else					_formErrors.append(tr("A control has no name"));
```
use addFormError in place of _formErrors.append: if not, no signal would be emitted.

JorisGoosen replied:
Oops!

boutinb reviewed (changes requested):
I have tested it, but it does not work well with malformed QML form.
For example if a control has no name, it generates an error, but the analysis is run, and clears the errors: you have no time to see the error.
Maybe a better solution would be to have a close icon (a cross) in the error message, so that the user can himself close the error message.

JorisGoosen commented:
Yeah it is a problem it clears the errors on refresh...
It would probably be better to do it the way you describe.

How would we clear the warning on an option then? 
As in, the errors on options are timed so solve themselves...

JorisGoosen committed:
Errors and warnings on a form are persistent but removable
- Moved the nice qml-cross from ControlErrorMessage to it's own qml file to reuse it
- removed the context "form" as it wasn't necessary
- processed some comments from PR
- Fixed a small bug where a control without name but with label would not be displayed with label.
- Controls without names now highlighted with warning color as well

JorisGoosen commented:
Ok so I've processed the requested changes and fixed a small bug.

Now form errors and warnings remain unless the user closes them.
Warnings/errors on options are not cleared but I think this will probably work out fine anyway.

The only warnings not set through known-issues are set because the "name" is missing or doubled or something. So they should stay. And if there are errors on options you can't run the analysis anyway.

boutinb commented on JASP-Desktop/analysis/analysisform.cpp:
```diff
@@ -721,49 +692,55 @@ void AnalysisForm::clearControlError(JASPControlBase* control)
+
+	_analysis = analysisPointer;
+
+	setAnalysisUp();
```
I don't see the need to call that here.
Why just to set _analysis here and let formCompleted to call the setup? This makes more complicated than it is.... Now you need this extra boolean _formCompleted.

JorisGoosen replied:
Because maybe `setAnalysis` is only set/called *after* `formCompleted`.
I don't see how right now, but I prefer to safeguard the code for it like this.

And I would also like to get rid of all that 'formCompleted' stuff at some point because I have the feeling it makes things more rigid than necessary. Because everything has to be set then and won't be changed later if users change properties or whatever.

boutinb reviewed (changes requested):
Just a small change. The rest seems to be perfect. It clears up the code!
